rngd.c

```diff
@@ -752,46 +752,80 @@ static error_t parse_opt (int key, char *arg, struct argp_state *state)
 static struct argp argp = { options, parse_opt, NULL, doc };
 
 
-static int update_kernel_random(struct rng *rng, int random_step,
-	unsigned char *buf, fips_ctx_t *fipsctx_in)
+static int update_kernel_random(int random_step)
 {
-	unsigned char *p;
-	int fips = 0;
+	unsigned char buf[FIPS_RNG_BUFFER_SIZE]; /* random_step was checked to be <= FIPS_RNG_BUFFER_SIZE */
 	int rc;
+	struct rng *iter;
 
-	if (!arguments->ignorefail)
-		fips = fips_run_rng_test(fipsctx_in, buf);
-	if (fips && !arguments->ignorefail)
-		return 1;
+	message(LOG_DAEMON|LOG_DEBUG, "entropy successfully gathered, preparing it for the kernel\n");
 
-	for (p = buf; p + random_step <= &buf[FIPS_RNG_BUFFER_SIZE];
-		 p += random_step) {
+	while(true)
+	{
 		if (!server_running)
 			return 0;
 		if (do_reseed) {
 			do_reseed = false;
 			alarm(arguments->force_reseed);
 		}
-		rc = random_add_entropy(p, random_step);
+
+		/* mix the sources on byte-level: ensure we always feed data from all available sources to the kernel
+		 * helps to mitigate problems should a source not be as random as expected */
+		int p = 0;
+		while(p < random_step) {
+			int progress = p;
+			for (int i = 0; i < ENT_MAX; ++i) {
+				iter = &entropy_sources[i];
+				if (!iter->entropy_buf.valid)
+					continue;
+
+				buf[p++] = iter->entropy_buf.entropy[iter->entropy_buf.used_pos++];
```

Comment:
given that this function only adds up to FIPS_RNG_BUFFER_SIZE of random data to the entropy pool, having each entropy source collect FIPS_RNG_BUFFER_SIZE worth of data, implies that we're going to discard FIPS_RNG_BUFFER_SIZE/ENT_MAX bytes from each entropy source on each iteration, which seems wasteful

Reply:
I'm not sure I fully understand your comment, especially the "FIPS_RNG_BUFFER_SIZE/ENT_MAX bytes" figure.

The intention of my code was to fill buf with random_step bytes (64 bytes by default) from all valid sources. then feed buf to the kernel.

After buf is fed to the kernel. the outer `while(true)` loop will continue. so if there are enough bytes left in the entropy_sources[].entropy_buf.entropy buffers they will be used. only when no source with data available is left, `p` will not progress anymore and the function will return with 0. then new entropy is gathered.

This is indeed not 100% using the available entropy, but only in the case when FIPS_RNG_BUFFER_SIZE is not an integer-multiple of random step. So in the default case of FIPS_RNG_BUFFER_SIZE = 2500 and random_step=64 we have 39 iterations (=2496 bytes) and 4 bytes left that are wasted per source. this is 0.16% which seems to be acceptable to me.

The alternative would be to decouple the buffers the data is gathered into and mixing the data. I think this will make the code more complicated and less easy to read and review. I think with a security sensitive program like rngd it is important that the code can be easily understood, reviewed and maintained.

Reply:
my concern was that we would stop when the kerenel entropy pool was full, but I guess the kernel entropy reporting is broken now anyway, so we just keep feeding until we decide we're done.  ok